Release v0.1

pranjalchanda08 opened the pull request.

pranjalchanda08 committed:
Workflow (#6)

* Create c-cpp.yml

* Update c-cpp.yml

pranjalchanda08 committed:
Human readable reg dump success

deveshshevde committed:
csr for review

deveshshevde committed:
csr for review

deveshshevde committed:
csr for review

deveshshevde committed:
csr for review

deveshshevde committed:
csr for review

deveshshevde committed:
csr for review

pranjalchanda08 committed:
Update LICENSE

deveshshevde committed:
added support for 25 csr instructions

deveshshevde committed:
added support for 25 csr instructions

pranjalchanda08 committed:
ELF loading added

pranjalchanda08 committed:
ELF loading added

deveshshevde committed:
added csrw csrr csrc support

deveshshevde committed:
added csrw csrr csrc support

deveshshevde committed:
added csrw csrr csrc support

deveshshevde committed:
added csrw csrr csrc support

akashkollipara commented on LICENSE:
```diff
@@ -19,3 +19,7 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 SOFTWARE.
+
+Note: Anyone Trying to reuse/replecate the codebase/code-architecure shall provide 
+viable credits to the author of the code. Any breach to the same would be subjected
```
~viable~ Due

akashkollipara commented on LICENSE:
```diff
@@ -5,7 +5,7 @@ Copyright (c) 2024 Pranjal Chanda
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
-to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+to use, modify, merge, publish, distribute, sublicense, and/or sell
```
allow only till modify

akashkollipara commented on LICENSE:
```diff
@@ -5,7 +5,7 @@ Copyright (c) 2024 Pranjal Chanda
 Permission is hereby granted, free of charge, to any person obtaining a copy
```
[x] Remove line 1. Rename the license

akashkollipara commented on inc/ram.h:
```diff
@@ -1,20 +1,17 @@
-#ifndef RAM_H
-#define RAM_H
+#ifndef _RAM_H_
```
Start using `#parma once`

akashkollipara commented on src/main.c:
```diff
@@ -1,17 +1,21 @@
+    ram = init_ram(RAM_SIZE);
+    if (ram)
+    {
+        printf("RAM Init done\n");
+    }
+    else
+    {
+        printf("[E: ] RAM Init Failed\n");
+        exit(EXIT_FAILURE);
```
```C
printf("%s\n", ram ? "RAM Initialized!":"[E: ] Failed to setup RAM");
return ram;
```

akashkollipara replied:
handle all error as part of exit routine where you de-init. Avoid exit.

akashkollipara replied:
https://github.com/VisorFolks/cc_elf_size/blob/bfe8a0e047e7c919de66e3a0921e9deeb9003ca4/elf_size.c#L272

akashkollipara commented on src/main.c:
```diff
@@ -1,17 +1,21 @@
-ram_t g_iram_mem;
+static rv_elf_section_info sections;
 
 static void rv32_ram_attach()
```
Return value.

pranjalchanda08 replied:
??

akashkollipara replied:
see next comment after this one.

akashkollipara commented:
use valgrind to check memory leaks during failure conditions.
most of the success will not have memory leaks but upon failure, there are.

shibcreate committed:
Validation activities

pranjalchanda08 committed:
jalr, store instruction bug fixes

shibcreate committed:
Updated JALR test to reflect correct execution

deveshshevde committed:
csr + jalr

deveshshevde committed:
csr + jalr

deveshshevde committed:
csr + jalr

deveshshevde committed:
csr + jalr

deveshshevde committed:
clean

deveshshevde committed:
clean

deveshshevde committed:
clean

deveshshevde committed:
clean

akashkollipara commented on inc/platmem.h:
```diff
@@ -1,14 +1,13 @@
 #define _K(_x)   (_x*1024)
 
+#ifndef RAM_SIZE
 #define RAM_SIZE  _K(16)
```
you dont need to make function like macro...

```
#define KB *1024
#define SIZE 16KB
```

This works too.

akashkollipara commented on inc/ram.h:
```diff
@@ -1,20 +1,17 @@
 #include "stdlib.h"
 
 typedef struct ram {
 	uint8_t *mem;     // ram memory of RAM_SIZE
```
RAM is word addressable not byte addressable.

pranjalchanda08 replied:
correct the load store is happening in word addressability only. Check `ram.c`

akashkollipara commented on inc/rv32I_struct.h:
```diff
@@ -139,6 +139,7 @@ typedef struct exec_args
     uint32_t fn7;
     ins_type_t ins_type;
     rv32i_ctx_t *c_ctx;
+    void * ram;
```
`uintptr_t *ram` would be better I suppose.

deveshshevde committed:
cleanup + fixes

deveshshevde committed:
cleanup + fixes

deveshshevde committed:
cleanup + fixes

deveshshevde committed:
cleanup + fixes

Commit:
Updated License

------ Summary ------
> Permitted
- Use
- Modify
- Merge
- Distribute
- Showcase work

> Cannot Claim
- Software Ownership
- Patent use or IP
- Warranty
- Liability

Commit:
Updated License

------ Summary ------
> Permitted
- Use
- Modify
- Merge
- Distribute
- Showcase work

> Cannot Claim
- Software Ownership
- Patent use or IP
- Warranty
- Liability

deveshshevde committed:
new header for csr

deveshshevde committed:
new header for csr

deveshshevde committed:
new header for csr

deveshshevde committed:
new header for csr

pranjalchanda08 committed:
added rv32_err_t for error manangement
inc directory revamped

pranjalchanda08 committed:
added rv32_err_t for error manangement
inc directory revamped

pranjalchanda08 committed:
added rv32_err_t for error manangement
inc directory revamped

pranjalchanda08 committed:
added rv32_err_t for error manangement
inc directory revamped

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
white space

deveshshevde committed:
allignment

deveshshevde committed:
allignment

shibcreate committed:
More validation .s files and rebased

shibcreate committed:
More validation scripts for RR --> All passing

pranjalchanda08 committed:
Documentation updates

deveshshevde committed:
csr reg-dump

deveshshevde committed:
csr reg-dump

deveshshevde committed:
compiler version independent

deveshshevde committed:
compiler version independent

deveshshevde committed:
Merge branch 'dev' into fr/4-add-csr-instruction-and-register-support

pranjalchanda08 committed:
logging framework addition

pranjalchanda08 committed:
Logging Fix

shibcreate committed:
Working signed validation cases

shibcreate committed:
Finishing validation for signed values

akashkollipara commented on Makefile:
```diff
@@ -1,69 +1,74 @@
-CXX = g++
-
+# Compiler settings
+CC := gcc
```
try using ccache for faster and efficient builds.
`CC := ccache gcc`
`CXX := ccache g++`

pranjalchanda08 replied:
this does not work
```sh
$ make
ccache gcc -Iinc/common -Iinc/tools -Iinc/device -Iinc/opcode -MMD -MP -g -c src/common/main.c -o build/src/common/main.c.o
make: ccache: Command not found
make: *** [Makefile:43: build/src/common/main.c.o] Error 127
```

akashkollipara commented on inc/common/main.h:
```diff
@@ -0,0 +1,25 @@
+ *****************************************************************************************/
+#pragma once
+
+#include "stdio.h"
+#include "string.h"
+#include "logging.h"
+#include "inttypes.h"
+#include "rv32_err.h"
+#include "ram.h"
+#include "elf_reader.h"
+#include "rv32I_struct.h"
+#include "rv32i_fetch_decode.h"
+#include "rv32i_execute.h"
```
are you sure you want to use `"..."` instead of `<...>`

pranjalchanda08 replied:
updated

akashkollipara commented on inc/common/rv32_err.h:
```diff
@@ -0,0 +1,20 @@
+
+typedef enum
+{
+    RV32_SUCCESS,
```
good practice is to assign at least first value...

akashkollipara commented on src/common/main.c:
```diff
@@ -0,0 +1,206 @@
+    memset(g_rv32i_ctx, 0, sizeof(rv32i_ctx_t));
+
+    err = rv32_ram_attach();
+    RV32_ASSERT_GOTO(err, exit);
+    /* Read instruction and data binary and save it to the ram instance */
+    err = rv32_ram_store_elf(argv[1], &entry_point);
+    RV32_ASSERT_GOTO(err, exit);
```
not all should goto exit...
it should to point to relevant deinit.

use pre_exit, exit, post_exit maybe?

akashkollipara commented on src/opcode/rv32_fetch_decode.c:
```diff
@@ -1,28 +1,131 @@
-     "s9", "s10", "s11", "t3", "t4", 
-     "t5", "t6"};
+const char *reg_name_list[] = {
+    "x0",
+    "ra",
+    "sp",
+    "gp",
+    "tp",
+    "t0",
+    "t1",
+    "t2",
+    "s0",
+    "s1",
+    "a0",
+    "a1",
+    "a2",
+    "a3",
+    "a4",
+    "a5",
+    "a6",
+    "a7",
+    "s2",
+    "s3",
+    "s4",
+    "s5",
+    "s6",
+    "s7",
+    "s8",
+    "s9",
+    "s10",
+    "s11",
+    "t3",
+    "t4",
+    "t5",
+    "t6",
+    "pc",
```
use real reg names instead of canonical names. or atleast create alias (macros or #defines).
btw... canonical name for `x0` is `zero`

pranjalchanda08 replied:
why shall we do that? These string are used to show the decoded instruction in the output. It is easier to read the alias names in that. One thing that can be done is to use the something like `r[n] <alias>` instead.

pranjalchanda08 replied:
```sh
-------------- Execution Start --------------
[22:38:35.003] [INFO ]: [0x00002117]: [PC:0x00001000]:  auipc   x2 <sp>   , 0x2000 <8192d>
[22:38:35.003] [INFO ]: [0x00010113]: [PC:0x00001004]:  addi    x2 <sp>   , x2 <sp>   , 0x00 <00d>
[22:38:35.003] [INFO ]: [0x01300293]: [PC:0x00001008]:  addi    x5 <t0>   , x0 <zero> , 0x13 <19d>
[22:38:35.003] [INFO ]: [0x02300313]: [PC:0x0000100C]:  addi    x6 <t1>   , x0 <zero> , 0x23 <35d>
[22:38:35.003] [INFO ]: [0x006283B3]: [PC:0x00001010]:  add     x7 <t2>   , x5 <t0>   , x6 <t1>   
[22:38:35.003] [INFO ]: [0x00001E97]: [PC:0x00001014]:  auipc   x29 <t4>  , 0x1000 <4096d>
[22:38:35.003] [INFO ]: [0xFECE8E93]: [PC:0x00001018]:  addi    x29 <t4>  , x29 <t4>  , 0xffffffec <-20d>
[22:38:35.003] [DEBUG]: PC reached EOF
-------------- Execution End ----------------
```

akashkollipara commented on src/opcode/rv32_fetch_decode.c:
```diff
@@ -1,28 +1,131 @@
+};
+
+static const char *csr_reg_list[] = {
+    "mstatus",
+    "misa",
+    "mie",
+    "mtvec",
+    "mtvt",
+    "mstatush",
+    "mcountinhibit",
+    "mcycle",
+    "mhpmevent31",
+    "mscratch",
+    "mepc",
+    "mcause",
+    "mtval",
+    "mip",
+    "mnxti",
+    "mintthresh",
+    "mscratchcswl",
+    "tselect",
+    "tdata1",
+    "tdata2",
+    "tinfo",
+    "dcsr",
+    "dpc",
+    "dscratch0",
+    "dscratch1",
+    "mcycle",
+    "minstret",
```
this might not be the right way to proceed as these registers do have non sequential addresses.

akashkollipara replied:
therefore by extension you might keep modifying this list. think of some alter approach. if the change is too big, let go and target it in upcoming release.

pranjalchanda08 replied:
we can revamp this. currently we can go ahead with the first release

pranjalchanda08 committed:
Comment fixes

shibcreate committed:
validation of store commands

pranjalchanda08 commented:
Waiting for @shibcreate to complete her L0 test on multiple instructions then we can plan on the release including those changes

pranjalchanda08 committed:
execute store bug fix

deveshshevde committed:
pr

deveshshevde committed:
build succe

deveshshevde committed:
CSR dump + Zicsr support

shibcreate committed:
Working validation of store instructions

deveshshevde committed:
comments for csr

shibcreate committed:
Final validation for last 3

shibcreate committed:
Created example README with changes

shibcreate committed:
Update README.md with 2 other examples

akashkollipara commented on Makefile:
```diff
@@ -1,69 +1,84 @@
 CC = gcc
 CXX = g++
+#Checks the compiler version and associate corresponding flag
+# eg : for [version > 11] _zicsr flag is required
+GCC_VERSION := $(shell riscv64-unknown-elf-gcc --version | head -n 1 | awk '{print $$3}')
+GCC_MAJOR := $(word 1, $(subst ., ,$(GCC_VERSION)))
+GCC_MINOR := $(word 2, $(subst ., ,$(GCC_VERSION)))
```
I am not happy with this implementation. But as this is not a blocker, I will ignore it.

Please try to refactor this change.

pranjalchanda08 replied:
I am closing all the Makefile related issue for now and creating a bug to fix it in next release.

akashkollipara commented on Makefile:
```diff
@@ -1,69 +1,84 @@
-# Note the single quotes around the * expressions. The shell will incorrectly expand these otherwise, but we want to send the * directly to the find command.
-SRCS := $(shell find $(SRC_DIRS) -name '*.cpp' -or -name '*.c' -or -name '*.s')
+# Locate all source files (C, C++, Assembly)
+SRCS := $(shell find $(SRC_DIRS) -type f \( -name '*.cpp' -o -name '*.c' -o -name '*.s' \))
```
## Fix needed
Doing this can be risky.
Consider this case, where one wants to create a c file and its not ready for build. The build tree will break.

akashkollipara commented on Makefile:
```diff
@@ -1,69 +1,84 @@
+# Compilation flags
 CPPFLAGS := $(INC_FLAGS) -MMD -MP
 
+ifeq ($(shell test $(GCC_MAJOR) -ge 11 && echo 1), 1)
+    MARCH_FLAG := -march=rv32i_zicsr
+else
+    MARCH_FLAG := -march=rv32i
+endif
+
```
Instead add min_version for newer flags.
```
if(version > min_verion)
 flag = abc;
else
 flag = def;
```

akashkollipara replied:
Avoid hard coding.

akashkollipara commented on Makefile:
```diff
@@ -1,69 +1,84 @@
+
 ASM ?=
-ARGS ?= 
+ARGS ?=
```
?= is used to assign if not assigned.

Alternatively dont use or assign a default value.

pranjalchanda08 replied:
The default value is supposed to be empty

akashkollipara commented on Makefile:
```diff
@@ -1,69 +1,84 @@
-run:
-	mkdir -p $(OUT_DIR)/$(ASM)
+# Run target
+.PHONY: run
```
phony is used when target doesnt exist and error will not be raised.

akashkollipara commented on rv32_asm/and.s:
```diff
@@ -0,0 +1,15 @@
+.globl stack_top
+
+_start:
+	la sp, stack_top
```
@pranjalchanda08 are you not mandating gp reg update to start the code?

pranjalchanda08 replied:
I am aware of this but currently GP init is not mandated. Shall be done once the exceptions are up.

akashkollipara reviewed (changes requested):
Please address the comments.

pranjalchanda08 commented:
@akashkollipara, all the issues are listed in https://github.com/pranjalchanda08/slash-sim/issues/33
Please approve for the release. 
We can fix it all in the upcoming release builds
